### ISSUE TITLE
can infer find out all suspecious bugs?

### ISSUE THREAD

**Issue by JiandongLu** [facebook/infer#1782]
currently I's using infer 1.1.0.
but i am curious the question : can infer find out all suspecious bugs?
if not, is there an open source  tool  can  do ?

thanks

**Comment by SolomonSun2010** [comment]
任何工具都不能找出所有的可疑bugs。例如，你如何定义C++、或Java/Android中的内存泄露问题，并让工具完全理解？同时，工具也要跟语言版本与时俱进。
除非你使用内存安全的语言，例如像Rust。

**Comment by hajduakos** [comment]
No, Infer is underapproximate by nature, which means that it tries to minimize false positives, while potentially having false negatives (missing bugs). [Software model checkers](https://sv-comp.sosy-lab.org/) aim to not have false negatives, however they often suffer from performance issues in practice.

**Comment by MPFuzz** [comment]
> No, Infer is underapproximate by nature, which means that it tries to minimize false positives, while potentially having false negatives (missing bugs). [Software model checkers](https://sv-comp.sosy-lab.org/) aim to not have false negatives, however they often suffer from performance issues in practice.

@hajduakos Where does the underapproximation in Infer's design originate: from its abstract interpretation or checker logic?

**Comment by hajduakos** [comment]
@MPFuzz to the best of my understanding, it originates from the join operation of the abstract domains defined by the checkers. For example, Pulse (one of the most used checkers) has a disjunctive domain, but only keeps a limited number of disjuncts. In other words, when multiple program points meet, Pulse keeps a limited number of the behaviors from each path instead of trying to merge them into a single (but overapproximatin) behavior.
